### Data Wrangling

In [195]:
import pandas as pd
import numpy as np
import  seaborn as sns

In [196]:
kashti=sns.load_dataset('titanic')
ks1=kashti
ks2=kashti


In [197]:
kashti.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [198]:
#simple operation on series like + - *
(kashti['age']+2).head()

0    24.0
1    40.0
2    28.0
3    37.0
4    37.0
Name: age, dtype: float64

### Dealing with missing values
in a dataset missing values are either ? or Nan \
jb data na ho ksi bhi 1 row me

## Steps
koshi kren dubara data collect kren  \
miising value wale column nikal den agre possible ho distubance na ho to \
replace  missing values

**Average**  
*frequency* or Mode \
*Replace on base of data* \
*Use ML algortihm* \
*Blank*


In [199]:
#see where missing values are
kashti.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [200]:
# use drop.na method
print(kashti.shape)
kashti.dropna(subset=['deck'], axis=0, inplace=True)  
#inplace= true modifies the dtaframe


(891, 15)


In [201]:
kashti.isnull().sum() #find agin null values

survived        0
pclass          0
sex             0
age            19
sibsp           0
parch           0
fare            0
embarked        2
class           0
who             0
adult_male      0
deck            0
embark_town     2
alive           0
alone           0
dtype: int64

In [202]:
kashti = kashti.dropna()# to drop na
# to update the main datafraame
kashti.isnull().sum() #removing ns from whole dataframe

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [203]:
kashti.shape

(182, 15)

In [204]:
ks1.isnull().sum()

survived        0
pclass          0
sex             0
age            19
sibsp           0
parch           0
fare            0
embarked        2
class           0
who             0
adult_male      0
deck            0
embark_town     2
alive           0
alone           0
dtype: int64

### Replacing missing values with the average of taht column

In [205]:
#finding an average (mean)
mean = ks1['age'].mean()
mean

35.77945652173913

In [206]:
# replacing nan with mean of the data (updating as well) 
ks1['age'] =  ks1 ['age'].replace(np.nan , mean)

In [207]:
ks1.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
deck           0
embark_town    2
alive          0
alone          0
dtype: int64

# Replcing missing values

In [208]:
mean = ks1['age'].mean()
mean

35.77945652173913

In [209]:
ks1['age']= ks1['age'].replace(np.nan ,mean)

In [210]:
ks1.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
deck           0
embark_town    2
alive          0
alone          0
dtype: int64

# Assignment
### Deck wali value ko mean k sath replace krnen "deck = 688"
### embark_town wali values ko remove kre

# Data Formatting 
### Data ko aik comman standard per lana
### Ensures data is consistent and understandable
### Easy to gather
### Easy to work with

In [211]:
#know the data type and convert ot into the known one
kashti.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [212]:
#use this method to convert datatypes from one to another format

kashti['survived'] = kashti ['survived'].astype("float64")
kashti.dtypes

C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/1043570814.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  kashti['survived'] = kashti ['survived'].astype("float64")


survived        float64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [213]:
#here we will conert age into days 
ks1["age"] = ks1["age"]*365
ks1.head(10)

survived  pclass     sex          age  sibsp  parch      fare embarked  \
1          1       1  female  13870.00000      1      0   71.2833        C   
3          1       1  female  12775.00000      1      0   53.1000        S   
6          0       1    male  19710.00000      0      0   51.8625        S   
10         1       3  female   1460.00000      1      1   16.7000        S   
11         1       1  female  21170.00000      0      0   26.5500        S   
21         1       2    male  12410.00000      0      0   13.0000        S   
23         1       1    male  10220.00000      0      0   35.5000        S   
27         0       1    male   6935.00000      3      2  263.0000        S   
31         1       1  female  13059.50163      1      0  146.5208        C   
52         1       1  female  17885.00000      1      0   76.7292        C   

     class    who  adult_male deck  embark_town alive  alone  
1    First  woman       False    C    Cherbourg   yes  False  
3    First  woman       False    C  Southampton   yes  False  
6    First    man        True    E  Southampton    no   True  
10   Third  child       False    G  Southampton   yes  False  
11   First  woman       False    C  Southampton   yes   True  
21  Second    man        True    D  Southampton   yes   True  
23   First    man        True    A  Southampton   yes   True  
27   First    man        True    C  Southampton    no  False  
31   First  woman       False    B    Cherbourg   yes  False  
52   First  woman       False    D    Cherbourg   yes  False

In [214]:
#Alwys rename afterwards
ks1.rename(columns = {"age": "age in days"}, inplace =True)
ks1.head()

survived  pclass     sex  age in days  sibsp  parch     fare embarked  \
1          1       1  female      13870.0      1      0  71.2833        C   
3          1       1  female      12775.0      1      0  53.1000        S   
6          0       1    male      19710.0      0      0  51.8625        S   
10         1       3  female       1460.0      1      1  16.7000        S   
11         1       1  female      21170.0      0      0  26.5500        S   

    class    who  adult_male deck  embark_town alive  alone  
1   First  woman       False    C    Cherbourg   yes  False  
3   First  woman       False    C  Southampton   yes  False  
6   First    man        True    E  Southampton    no   True  
10  Third  child       False    G  Southampton   yes  False  
11  First  woman       False    C  Southampton   yes   True

# Data Normalization
### Uniform the data
### Making user they have same impect
### aik samneder me ak jar me
### Also for compututional reasons

In [215]:
ks1.head()

survived  pclass     sex  age in days  sibsp  parch     fare embarked  \
1          1       1  female      13870.0      1      0  71.2833        C   
3          1       1  female      12775.0      1      0  53.1000        S   
6          0       1    male      19710.0      0      0  51.8625        S   
10         1       3  female       1460.0      1      1  16.7000        S   
11         1       1  female      21170.0      0      0  26.5500        S   

    class    who  adult_male deck  embark_town alive  alone  
1   First  woman       False    C    Cherbourg   yes  False  
3   First  woman       False    C  Southampton   yes  False  
6   First    man        True    E  Southampton    no   True  
10  Third  child       False    G  Southampton   yes  False  
11  First  woman       False    C  Southampton   yes   True

In [216]:
ks4= ks1[["age in days", 'fare']]
ks4.head(10)

age in days      fare
1   13870.00000   71.2833
3   12775.00000   53.1000
6   19710.00000   51.8625
10   1460.00000   16.7000
11  21170.00000   26.5500
21  12410.00000   13.0000
23  10220.00000   35.5000
27   6935.00000  263.0000
31  13059.50163  146.5208
52  17885.00000   76.7292

In [217]:
kashti.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1        1.0       1  female  38.0      1      0  71.2833        C  First   
3        1.0       1  female  35.0      1      0  53.1000        S  First   
6        0.0       1    male  54.0      0      0  51.8625        S  First   
10       1.0       3  female   4.0      1      1  16.7000        S  Third   
11       1.0       1  female  58.0      0      0  26.5500        S  First   

      who  adult_male deck  embark_town alive  alone  
1   woman       False    C    Cherbourg   yes  False  
3   woman       False    C  Southampton   yes  False  
6     man        True    E  Southampton    no   True  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True

# Normalization

In [218]:
ks4 ['fare'] = ks4['fare'] /ks4['fare'].max()
ks4 ['age in days'] = ks4['age in days'] /ks4['age in days'].max()
ks4.head()

C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/603556916.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks4 ['fare'] = ks4['fare'] /ks4['fare'].max()
C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/603556916.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks4 ['age in days'] = ks4['age in days'] /ks4['age in days'].max()


age in days      fare
1        0.4750  0.139136
3        0.4375  0.103644
6        0.6750  0.101229
10       0.0500  0.032596
11       0.7250  0.051822

In [188]:
#min - Max Method
ks4['fare'] = (ks4['fare']-ks4['fare'].min())/ks4['fare'].max()-ks4['fare'].min()
ks4.head()

C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/3606290560.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks4['fare'] = (ks4['fare']-ks4['fare'].min())/ks4['fare'].max()-ks4['fare'].min()


age in days      fare
1        0.4750  0.139136
3        0.4375  0.103644
6        0.6750  0.101229
10       0.0500  0.032596
11       0.7250  0.051822

In [189]:
#Z-score (standard score)
ks4['fare'] = (ks4['fare']-ks4['fare'].mean())/ks4['fare'].std()
ks4.head()

C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/1867067840.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks4['fare'] = (ks4['fare']-ks4['fare'].mean())/ks4['fare'].std()


age in days      fare
1        0.4750 -0.067879
3        0.4375 -0.311883
6        0.6750 -0.328489
10       0.0500 -0.800339
11       0.7250 -0.668161

In [190]:
# # Log transformation
# ks['fare'] = np.log(ks['fare'])
# ks.head()
# Baer y data pr krna chaie normalization

### Binning
Groupping of values into smaller 
convert numeric into categories

In [219]:
bins=np.linspace(min(kashti['age']),max(kashti['age']), 15000)
age_groups=["bachey", "jawan", "boorhey"]
kashti['age']= pd.cut(kashti['age'], bins, include_lowest=True)
labels=['10-20','21-30', '40-60']
kashti['age']
#Assignment


C:\Users\Faiza\AppData\Local\Temp/ipykernel_11196/2090269086.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  kashti['age']= pd.cut(kashti['age'], bins, include_lowest=True)


1        (37.995, 38.0]
3        (34.995, 35.0]
6      (53.997, 54.002]
10       (3.999, 4.004]
11     (57.998, 58.004]
             ...       
871      (46.995, 47.0]
872    (32.997, 33.002]
879      (55.995, 56.0]
887    (18.999, 19.004]
889    (25.995, 26.001]
Name: age, Length: 182, dtype: category
Categories (14999, interval[float64, right]): [(0.919, 0.925] < (0.925, 0.931] < (0.931, 0.936] < (0.936, 0.941] ... (79.979, 79.984] < (79.984, 79.989] < (79.989, 79.995] < (79.995, 80.0]]

In [192]:
#Assignment 3 age group 
# how this will change the names in dataset based on grouping(Assignment)

###** Converting categories into dumies
easy to use for computation
Male Female(0,1)

In [193]:
pd.get_dummies(ks1['sex'])#Append dummies values
#How to use get dumies to change data inside a dataframe(Assgnemnt)

female  male
1         1     0
3         1     0
6         0     1
10        1     0
11        1     0
..      ...   ...
871       1     0
872       0     1
879       1     0
887       1     0
889       0     1

[203 rows x 2 columns]

In [224]:
#transfer into dummy values
ks1['sex']= ks1.append(['sex'])

ValueError: Columns must be same length as key